OpenRA.Mods.Common/Traits/World/TerrainGeometryOverlay.cs

```diff
@@ -73,54 +37,34 @@ public void Render(WorldRenderer wr)
 			if (devMode.Value == null || !devMode.Value.ShowTerrainGeometry)
 				return;
 
-			var ts = wr.World.Map.TileShape;
+			var map = wr.World.Map;
+			var tileSet = wr.World.TileSet;
+			var lr = Game.Renderer.WorldLineRenderer;
 			var colors = wr.World.TileSet.HeightDebugColors;
-
-			var leftDelta = ts == TileShape.Diamond ? new WVec(-512, 0, 0) : new WVec(-512, -512, 0);
-			var topDelta = ts == TileShape.Diamond ? new WVec(0, -512, 0) : new WVec(512, -512, 0);
-			var rightDelta = ts == TileShape.Diamond ? new WVec(512, 0, 0) : new WVec(512, 512, 0);
-			var bottomDelta = ts == TileShape.Diamond ? new WVec(0, 512, 0) : new WVec(-512, 512, 0);
+			var mouseCell = wr.Viewport.ViewToWorld(Viewport.LastMousePos).ToMPos(wr.World.Map);
 
 			foreach (var uv in wr.Viewport.VisibleCells.MapCoords)
 			{
-				var lr = Game.Renderer.WorldLineRenderer;
-				var pos = wr.World.Map.CenterOfCell(uv.ToCPos(wr.World.Map));
+				var height = (int)map.MapHeight.Value[uv];
+				var tile = map.MapTiles.Value[uv];
+				var ti = tileSet.GetTileInfo(tile);
+				var ramp = ti != null ? ti.RampType : 0;
 
-				var height = (int)wr.World.Map.MapHeight.Value[uv];
-				var tile = wr.World.Map.MapTiles.Value[uv];
+				var corners = map.CellCorners[ramp];
+				var color = corners.Select(c => colors[height + c.Z / 512]).ToArray();
+				var pos = map.CenterOfCell(uv.ToCPos(map));
+				var screen = corners.Select(c => wr.ScreenPxPosition(pos + c).ToFloat2()).ToArray();
 
-				TerrainTileInfo tileInfo = null;
+				if (uv == mouseCell)
+					lr.LineWidth = 3;
 
-				// TODO: This is a temporary workaround for our sloppy tileset definitions
-				// (ra/td templates omit Clear tiles from templates)
-				try
+				for (var i = 0; i < 4; i++)
 				{
-					tileInfo = wr.World.TileSet.Templates[tile.Type][tile.Index];
+					var j = (i + 1) % 4;
```

Comment:
This is equivalent to:
```csharp
var j = i + 1;
while (j > 4)
	j -= 4;
```
I don't think it's appropriate to comment standard mathematical operations.